sdk/client/src/stream/websocket_transport.rs

```diff
@@ -0,0 +1,156 @@
+        let id = self.next_id.fetch_add(1, Ordering::Relaxed);
+        Id::Number(id)
+    }
+
+    /// This consumes the websocket stream (but not the sink)
+    /// Possibilities:
+    /// 1. `stream.next().await` returns `None`: this happens if a stream is closed
+    /// 2. `stream.next().await` returns `Some(Err)`: this is generally a transport error, and the
+    ///     stream can be considered closed
+    /// 3. `stream.next().await` returns `Some(Ok(Message))`: this may or may not be valid JSON from
+    ///     the server, so we need to do a bit of validation:
+    ///     1. `msg.is_text()` must be `true`: we ignore `Ping`/`Pong`, `Binary`, etc message types
+    ///     2. `msg.to_text()` must be `Ok(String)`: Ensures the message is valid UTF8
+    ///     3. `StreamJsonRpcResponse::from_str(msg)` must be `Ok(StreamJsonRpcResponse)`: this is
+    ///         serde deserialization, which validates that the JSON is well formed, and matches the
+    ///         `StreamJsonRpcResponse` struct
+    ///
+    pub fn get_stream(mut self) -> (StreamingClientReceiver, Self) {
+        let (sender, receiver) = mpsc::channel(100);
+
+        let mut stream = self
+            .stream
+            .expect("Stream is `None`: it has already been consumed");
+        self.stream = None;
+
+        self.channel_task = Some(tokio::task::spawn(async move {
+            loop {
+                match stream.next().await {
+                    None => {
+                        sender
+                            .send(Err(StreamError::connection_closed(None::<StreamError>)))
+                            .await
+                            .ok();
+                        continue;
+                    }
+                    Some(msg) => match msg {
+                        Ok(msg) => {
+                            if msg.is_text() {
+                                let msg = match msg
+                                    .to_text()
+                                    .map_err(StreamError::from_tungstenite_error)
+                                {
+                                    Ok(msg) => msg,
+                                    Err(e) => {
+                                        let _ = sender.send(Err(e)).await;
+                                        continue;
+                                    }
+                                };
+                                match StreamJsonRpcResponse::from_str(msg) {
+                                    Ok(msg) => sender.send(Ok(msg)).await.ok(),
+                                    Err(e) => sender.send(Err(StreamError::from(e))).await.ok(),
+                                };
+                            }
+                        }
+                        Err(e) => {
+                            let _ = sender
+                                .send(Err(StreamError::from_tungstenite_error(e)))
+                                .await;
+                            continue;
```

Comment:
same

Reply:
when there is an upstream error, it sends a message to the client so that the client can handle the shutdown in a nice way on their end